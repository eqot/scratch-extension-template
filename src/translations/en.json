{
  "say": "say [MESSAGE]",

  "showSprite": "show [TARGET] in log",
  "all sprites": "all sprites",
  "this sprite": "this sprite"
}
