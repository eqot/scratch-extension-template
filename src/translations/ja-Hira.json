{
  "say": "[MESSAGE]という",

  "showSprite": "ログに[TARGET]をひょうじする",
  "all sprites": "すべてのスプライト",
  "this sprite": "このスプライト"
}
