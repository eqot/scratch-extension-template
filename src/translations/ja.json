{
  "say": "[MESSAGE]と言う",

  "showSprite": "ログに[TARGET]を表示する",
  "all sprites": "すべてのスプライト",
  "this sprite": "このスプライト"
}
