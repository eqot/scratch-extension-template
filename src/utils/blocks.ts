import { Block } from '../types';

const Separator = '---';

export const generateBlocks = (blocksOrder: string[], object: any): any => {
  const blocks = blocksOrder.map((block: any) => {
    if (isSeparator(block)) {
      return block;
    }

    const { info, menus, ...functions } = require(`../blocks/${block}.ts`).default;
    return { info, menus, functions };
  }) as Block[];

  const info = (): any => blocks.map((block) => (isSeparator(block) ? block : block.info()));
  const menus = (): any =>
    blocks.reduce((acc, { menus }) => (menus ? Object.assign(acc, menus()) : acc), {});
  const functions: any = blocks.reduce((acc, { functions }) => Object.assign(acc, functions), {});

  for (const functionName in functions) {
    object[functionName] = functions[functionName].bind(object);
  }

  return {
    info,
    menus,
    functions,
  };
};

function isSeparator(block: unknown): boolean {
  return block === Separator;
}
