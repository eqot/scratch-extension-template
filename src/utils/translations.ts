import formatMessage from 'format-message';

const SUPPORTED_LOCALES = ['en', 'ja', 'ja-Hira'];

class Translations {
  initialize(defaultLocale?: string, runtime?: any): void {
    formatMessage.setup({
      locale: defaultLocale || window.navigator.language || 'en',
      translations: SUPPORTED_LOCALES.reduce(
        (acc, locale) =>
          Object.assign(acc, { [locale]: require(`../translations/${locale}.json`) }),
        {}
      ),
    });

    if (runtime) {
      runtime.on('LOCALE_CHANGED', (locale: string) => this.setLocale(locale));
    }
  }

  setLocale(locale: string): void {
    if (locale !== formatMessage.setup().locale) {
      formatMessage.setup({ locale });
    }
  }

  label(id: string): string {
    return formatMessage(id);
  }
}

const translations = new Translations();

export { translations };
