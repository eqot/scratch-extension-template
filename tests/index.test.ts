import Runtime from 'scratch-vm/src/engine/runtime.js';

import DummyExtensions from '../src/index';
import { translations } from '../src/utils/translations';

describe('DummyBlocks', () => {
  let extension: DummyExtensions;

  beforeEach(() => {
    const runtime = new Runtime();
    extension = new DummyExtensions(runtime);
  });

  it('returns block information as expected in English', () => {
    translations.setLocale('en');

    const result = extension.getInfo();
    expect(result).toMatchSnapshot();
  });

  it('returns block information as expected in Japanese', () => {
    translations.setLocale('ja');

    const result = extension.getInfo();
    expect(result).toMatchSnapshot();
  });
});
