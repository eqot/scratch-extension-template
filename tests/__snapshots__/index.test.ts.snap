// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DummyBlocks returns block information as expected in English 1`] = `
Object {
  "blockIconURI": "",
  "blocks": Array [
    Object {
      "arguments": Object {
        "MESSAGE": Object {
          "defaultValue": "Hello, World!",
          "type": "string",
        },
      },
      "blockType": "reporter",
      "opcode": "say",
      "text": "say [MESSAGE]",
    },
    "---",
    Object {
      "arguments": Object {
        "TARGET": Object {
          "menu": "Target",
          "type": "string",
        },
      },
      "blockType": "command",
      "opcode": "showSprite",
      "text": "show [TARGET] in log",
    },
  ],
  "color1": "#a0a0a0",
  "color2": "#808080",
  "color3": "#606060",
  "id": "dummy",
  "menuIconURI": "",
  "menus": Object {
    "Target": Object {
      "acceptReporters": true,
      "items": Array [
        Object {
          "text": "all sprites",
          "value": "allSprites",
        },
        Object {
          "text": "this sprite",
          "value": "thisSprite",
        },
      ],
    },
  },
  "name": "Dummy",
}
`;

exports[`DummyBlocks returns block information as expected in Japanese 1`] = `
Object {
  "blockIconURI": "",
  "blocks": Array [
    Object {
      "arguments": Object {
        "MESSAGE": Object {
          "defaultValue": "Hello, World!",
          "type": "string",
        },
      },
      "blockType": "reporter",
      "opcode": "say",
      "text": "[MESSAGE]と言う",
    },
    "---",
    Object {
      "arguments": Object {
        "TARGET": Object {
          "menu": "Target",
          "type": "string",
        },
      },
      "blockType": "command",
      "opcode": "showSprite",
      "text": "ログに[TARGET]を表示する",
    },
  ],
  "color1": "#a0a0a0",
  "color2": "#808080",
  "color3": "#606060",
  "id": "dummy",
  "menuIconURI": "",
  "menus": Object {
    "Target": Object {
      "acceptReporters": true,
      "items": Array [
        Object {
          "text": "すべてのスプライト",
          "value": "allSprites",
        },
        Object {
          "text": "このスプライト",
          "value": "thisSprite",
        },
      ],
    },
  },
  "name": "Dummy",
}
`;
